src/Compilers/Core/AnalyzerDriver/AnalyzerManager.AnalyzerAndOptions.cs

```diff
@@ -18,7 +18,13 @@ public AnalyzerAndOptions(DiagnosticAnalyzer analyzer, AnalyzerOptions analyzerO
                 Debug.Assert(analyzerOptions != null);
 
                 Analyzer = analyzer;
-                _analyzerOptions = analyzerOptions;
+                // _analyzerOptions is used only as a key for equality testing, and only the properties
+                // declared by AnalyzerOptions are appropriate for use in the comparison. (For example,
+                // two AnalyzersAndOptions that differ only by Workspace ought to be treated as equal.)
+                // If the supplied analyzerOptions is an instance of a derived type, holding on to that object
+                // can introduce a memory leak, as well as causing false negatives in comparison. Avoid these
+                // issues by creating a new object with only the desired properties. 
+                _analyzerOptions = new AnalyzerOptions(analyzerOptions.AdditionalFiles);
```

Comment:
John, you may also want to add a note that this a temporary fix. Once #2706 is fixed, we can get rid of `WorkspaceAnalyzerOptions` hack and seal `AnalyzerOptions`.